### ISSUE TITLE
setInterval keeps on posting updates after the account has been switched

### ISSUE THREAD

**Issue by tiabc** [status-im/status-go#260]
@rasom:
>i used this code https://gist.github.com/rasom/a9b03b83abc9d3345fb8671f21d6d9be
so when user opens chat `init` function is called and every 3 seconds a new message appears. The problem is that when we login to another account bot’s vm is still running and messages are sent to this another account

The reason:
@adambabik:
> https://github.com/status-im/status-go/blob/develop/geth/jail/jail.go#L163-L170

The point is that we should ensure we destroy all jail cells upon account switching and stop all ottoext loops.

**Comment by tiabc** [comment]
@oleg-raev is looking into that. Please, don't take this issue.

**Comment by tiabc** [comment]
#216 seems a blocker here as I currently see no way of terminating an already running ottoext.Loop in its api.

**Comment by oleg-raev** [comment]
As commented @divan for https://github.com/status-im/status-go/issues/216#issuecomment-326497825
> I believe this issue was a part of #283 and can be closed.

@tiabc 
Is it means that now this issue blocked by the issue #283 ?

**Comment by tiabc** [comment]
@divan why? There's still no cleanup after `SwitchAccount`, right?
@oleg-raev it's either not blocked or not actual any more. Let's wait for @divan's reply.

By the way, regarding your architecture outline. The whole jail looks like it's dependent on the current, say, execution context where there's at least an injected account and an associated jail.
Not sure that execution context is the best naming but it perfectly conveys the essence.

**Comment by divan** [comment]
@tiabc my comment was related to #216 only, which doesn't say anything about switching account, but mainly about implementation of single goroutine with loop per jail.

**Comment by tiabc** [comment]
Thanks, @divan.
@oleg-raev the status of this issue is up-to-date, you can freely work on it.

**Comment by oleg-raev** [comment]
Thanks, @tiabc 
I have a question. Is there any way to stop the loop from AccountManager context (https://github.com/status-im/status-go/blob/develop/geth/node/accounts.go#L198)?

**Comment by tiabc** [comment]
@oleg-raev I don't think AccountManager is the right place to terminate JailCell's loop as they should be two orthogonal concepts. It's better be done on a higher level.

**Comment by divan** [comment]
>  Is there any way to stop the JailCell's loop from AccountManager context 
Currently, there is no way to stop loop at all (and that's kinda ok for now), but we were thinking about introducing this via context.Context and cancelling when user is logging out.

If logging out is only initiated via API, we can add "cancel jail's context" from upper API level, which has access to JailManager:
https://github.com/status-im/status-go/blob/develop/geth/api/api.go#L153

**Comment by deoxxa** [comment]
Hello! I see you're having some issues with ottoext. Anything I can do to help?

**Comment by divan** [comment]
Hi @deoxxa! Thanks for offering your time, we actually realized that we have a bit different constrains for event loop, so we forked `ottoext` and are using it internally now: https://github.com/status-im/status-go/tree/develop/geth/jail/internal

**Comment by deoxxa** [comment]
Oh, interesting! What are the different constraints you have? If they're applicable in general I'd be happy to look at merging some changes back upstream - if you're okay with that :)

**Comment by divan** [comment]
We'd be happy to contribute, of course. In our case event loop is a long running - actually it's "forever" running thing. Async requests may come at any time and they should be executed in the loop context.

So there is no such thing as "closed loop" in our case, for example.

Another major thing we've hit is that otto vm is not safe for concurrent calls. In many cases when even loop was involved it was panicking with "concurrent map read and write" error. I think this panic for unsafe map access has been added in Go 1.8, fairly recently.

Basically, our changes are fixes around these two things. Plus smaller things like using the same test framework as the rest of the code, etc.

**Comment by deoxxa** [comment]
Ah, I see.

The long-running loop workload is actually one of the things ottoext can really help with - as long as you run all tasks via the event loop (via `EvalTask` or `CallTask`) and not direct `Run` or `Exec` calls on the underlying VM. Indeed this is how I use it in several projects, with ~100 goroutines all running tasks in turn.

As for otto itself, yeah, it has about the same guarantees as a `map` does. You can probably do read-only stuff (fetching/converting values) from multiple goroutines, but not anything that modifies the state of the VM (e.g. running code or setting globals). But again, that's one of the things I built ottoext to address - sharing a single JavaScript execution context between multiple goroutines.

I'll take a poke through the code here to see how you're using the loop interfaces - I'm very curious about how other people might be using my little project! If I find anything that could be improved, I'll let you know.

**Comment by divan** [comment]
Cool, thanks! Feel free to ask any question on our use case. One particular constraint is that we don't have control over the JS code being executed - it's user supplied code in some scenarios.

To help you a bit - we're using loop with `timers`/`fetch` callbacks, and `CallTask` for some asynchronous user methods.

**Comment by tiabc** [comment]
@deoxxa thanks for getting back, we really appreciate that! Feel free to open a new issue here to discuss your thoughts or we you can do it on our slack channel: http://slack.status.im/ pinging @divan

Closing the issue as it's been resolved.
